File: app/api/v1/knowledge.py
"""Knowledge-base endpoints — document CRUD + retrieval debug (RAG, priority 57).

Documents are tenant-scoped: a store only ever sees its own knowledge base
(enforced at the repository layer). Permission granularity mirrors the unified
permission model: ``knowledge:read`` (GET) seeded for owner/admin/member,
``knowledge:create`` / ``knowledge:delete`` seeded for owner/admin (admin has
no delete). Super admins short-circuit via ``permission_service.check``.

Create ingests inline (split + embed + index) within the request — MVP scope,
no background task. If the embedding provider is misconfigured the document is
still created but marked ``status=failed`` so the admin sees it in the list.
"""

from fastapi import APIRouter, Depends, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import CurrentUser, get_current_user, require_permission
from app.core.database import get_db
from app.schemas.document import (
    DistributeRequest,
    DocumentCreate,
    DocumentRead,
    KnowledgeCategoryCreate,
    KnowledgeCategoryRead,
    KnowledgeCategoryUpdate,
    KnowledgeDistributionRead,
    RetrieveRequest,
    RetrieveResult,
)
from app.services.category_service import KnowledgeCategoryService
from app.services.knowledge_service import KnowledgeService

router = APIRouter(prefix="/knowledge", tags=["knowledge"])


@router.get(
    "/documents",
    response_model=list[DocumentRead],
    dependencies=[Depends(require_permission("knowledge", "read"))],
)
async def list_documents(
    user: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> list[DocumentRead]:
    """List the caller's tenant documents."""
    return await KnowledgeService(db).list_documents(
        user.user_id, user.tenant_id, platform_role=user.platform_role
    )


@router.post(
    "/documents",
    response_model=DocumentRead,
    status_code=status.HTTP_201_CREATED,
    dependencies=[Depends(require_permission("knowledge", "create"))],
)
async def create_document(
    payload: DocumentCreate,
    user: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> DocumentRead:
    """Create a document and ingest it (split + embed + index) inline."""
    return await KnowledgeService(db).create_document(
        user.user_id,
        user.tenant_id,
        payload,
        platform_role=user.platform_role,
    )


@router.delete(
    "/documents/{document_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    dependencies=[Depends(require_permission("knowledge", "delete"))],
)
async def delete_document(
    document_id: str,
    user: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> None:
    """Soft-delete a document and drop its chunks."""
    await KnowledgeService(db).delete_document(
        user.user_id,
        user.tenant_id,
        document_id,
        platform_role=user.platform_role,
    )


@router.post(
    "/retrieve",
    response_model=RetrieveResult,
    dependencies=[Depends(require_permission("knowledge", "read"))],
)
async def retrieve(
    payload: RetrieveRequest,
    user: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> RetrieveResult:
    """Debug retrieval: embed the query, return top-k chunks with scores.

    Lets an admin verify the RAG pipeline finds the right context before
    relying on it in agent conversations. Requires Postgres (pgvector) — on
    SQLite this is exercised via mocked embeddings.
    """
    return await KnowledgeService(db).retrieve_for_debug(
        user.user_id,
        user.tenant_id,
        payload.query,
        top_k=payload.top_k,
        platform_role=user.platform_role,
    )


# --------------------------------------------------- distribution (slice 03)
# D3 explicit distribution: a superior pushes a source document to target store(s)
# by writing knowledge_distribution rows (the reference-model link, D4). Two
# endpoints — POST to distribute (G4: target_tenant_ids XOR target_group_id),
# DELETE to revoke (soft flip is_active=false, audit preserved). Both require the
# new ``knowledge:distribute`` perm (seeded for owner/admin, not member; the
# group_admin bypass fires via the service's require(db=self.db)).


@router.post(
    "/documents/{document_id}/distribute",
    response_model=list[KnowledgeDistributionRead],
    status_code=status.HTTP_201_CREATED,
    dependencies=[Depends(require_permission("knowledge", "distribute"))],
)
async def distribute_document(
    document_id: str,
    payload: DistributeRequest,
    user: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> list[KnowledgeDistributionRead]:
    """Push a source document to target store(s).

    Body is G4 XOR: ``target_tenant_ids`` (explicit list) OR ``target_group_id``
    (expand to every store in that group). group_admin may only target their own
    group; super_admin is unrestricted. Returns the resulting distribution rows.
    """
    return await KnowledgeService(db).distribute_document(
        user.user_id,
        user.tenant_id,
        document_id,
        payload,
        platform_role=user.platform_role,
    )


@router.delete(
    "/distributions/{distribution_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    dependencies=[Depends(require_permission("knowledge", "distribute"))],
)
async def revoke_distribution(
    distribution_id: str,
    user: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> None:
    """Soft-revoke a distribution (is_active=false; audit row preserved).

    After revoke the target store's list/retrieve automatically exclude the doc
    (their ``dist.is_active=True`` predicate drops it) — no manual flip needed.
    """
    await KnowledgeService(db).revoke_distribution(
        user.user_id,
        user.tenant_id,
        distribution_id,
        platform_role=user.platform_role,
    )


# -------------------------------------------------------- categories (slice 01)
# Tiered Category CRUD (plan-knowledge-tiered-backend slice 01). The casbin
# gate reuses the existing knowledge:read/create/delete codes (G6: no new code);
# KnowledgeCategoryService adds the scope↔role binding (platform→super_admin /
# group→group_admin / store→owner-admin) on top. ``update`` rides the
# ``knowledge:create`` write gate (no knowledge:update code exists — documents
# have no edit path; see KnowledgeCategoryService.update docstring).


@router.get(
    "/categories",
    response_model=list[KnowledgeCategoryRead],
    dependencies=[Depends(require_permission("knowledge", "read"))],
)
async def list_categories(
    user: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> list[KnowledgeCategoryRead]:
    """List Categories visible to the caller (platform + own group + own store).

    group_admin gets the aggregated chain view (all sibling stores in their
    group); super_admin/hq_staff see everything.
    """
    return await KnowledgeCategoryService(db).list(
        user.user_id, user.tenant_id, platform_role=user.platform_role
    )


@router.post(
    "/categories",
    response_model=KnowledgeCategoryRead,
    status_code=status.HTTP_201_CREATED,
    dependencies=[Depends(require_permission("knowledge", "create"))],
)
async def create_category(
    payload: KnowledgeCategoryCreate,
    user: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> KnowledgeCategoryRead:
    """Create a Category in the caller's authorized scope (G6 scope↔role check)."""
    return await KnowledgeCategoryService(db).create(
        user.user_id, user.tenant_id, payload, platform_role=user.platform_role
    )


@router.put(
    "/categories/{category_id}",
    response_model=KnowledgeCategoryRead,
    dependencies=[Depends(require_permission("knowledge", "create"))],
)
async def update_category(
    category_id: str,
    payload: KnowledgeCategoryUpdate,
    user: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> KnowledgeCategoryRead:
    """Update a Category's name/sort_order (scope and ownership are immutable)."""
    return await KnowledgeCategoryService(db).update(
        user.user_id, user.tenant_id, category_id, payload,
        platform_role=user.platform_role,
    )


@router.delete(
    "/categories/{category_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    dependencies=[Depends(require_permission("knowledge", "delete"))],
)
async def delete_category(
    category_id: str,
    user: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> None:
    """Soft-delete a Category (is_deleted=true; name reusable afterwards)."""
    await KnowledgeCategoryService(db).delete(
        user.user_id, user.tenant_id, category_id,
        platform_role=user.platform_role,
    )

File: frontend/src/hooks/queries/knowledge.ts
/**
 * queries/knowledge — knowledge base / RAG (priority 57).
 *
 * Extracted from queries.ts monolith (plan-queries-endpoints-domain-split.md).
 * Locality move; see ./core.ts for qk + useApiMutation.
 *
 * knowledge-tiered reader-ui slice 01 G6 — ``useDocuments`` accepts the
 * optional {scope, category_id} filter and forwards it to fetchDocuments;
 * the query key encodes the filter so each distinct filter caches
 * independently. New ``useKnowledgeCategories`` powers the left-pane
 * category-tree (slice 02).
 */
import { qk, useApiMutation } from "./core";
import { useQuery } from "@tanstack/react-query";
import {
  createCategory,
  createDocument,
  deleteCategory,
  deleteDocument,
  distributeDocument,
  fetchDocuments,
  fetchKnowledgeCategories,
  listDistributions,
  revokeDistribution,
  updateCategory,
  type DocumentListFilter,
} from "@/api/endpoints";
import type {
  DistributeRequest,
  DocumentCreate,
  KnowledgeCategoryCreate,
  KnowledgeCategoryUpdate,
} from "@/api/types";
// ---------- knowledge base / RAG (priority 57) ----------
// knowledge-tiered reader-ui slice 01 G6: filter forwarded to fetchDocuments;
// query key encodes the filter so each distinct (scope, category_id) cell
// caches independently (mirrors qk.conversations' filter-encoding pattern).
export function useDocuments(filter?: DocumentListFilter) {
  return useQuery({
    queryKey: qk.documents(filter),
    queryFn: () => fetchDocuments(filter),
  });
}

// knowledge-tiered reader-ui slice 01 G6 — categories for the left-pane tree.
export function useKnowledgeCategories() {
  return useQuery({
    queryKey: qk.knowledgeCategories,
    queryFn: fetchKnowledgeCategories,
  });
}

export function useCreateDocument() {
  return useApiMutation(
    (payload: DocumentCreate) => createDocument(payload),
    [qk.documents()],
  );
}

export function useDeleteDocument() {
  return useApiMutation((id: string) => deleteDocument(id), [qk.documents()]);
}

// ---------- distributions (admin-ui slice 02 / backend slice 03) ----------
// knowledge-tiered admin-ui:下发/撤回/list hooks。useDistributions 按 docId 缓存;
// 下发/撤回成功后失效该 doc 的 distribution 列表 + 文档列表(下发不新增文档,但撤回
// 影响门店侧可见性,保守失效 documents 一并刷新)。

/** 列出某文档的所有下发关系(含已撤回 is_active=false)。供「管理下发」视图。 */
export function useDistributions(docId: string) {
  return useQuery({
    queryKey: qk.documentDistributions(docId),
    queryFn: () => listDistributions(docId),
  });
}

/** 下发文档到指定门店(target_tenant_ids)或整个集团(target_group_id),XOR。 */
export function useDistributeDocument(docId: string) {
  return useApiMutation(
    (payload: DistributeRequest) => distributeDocument(docId, payload),
    [qk.documentDistributions(docId)],
  );
}

/** 撤回一条下发(软删 is_active=false)。失效该 doc 的 distribution 列表。 */
export function useRevokeDistribution(docId: string) {
  return useApiMutation(
    (distId: string) => revokeDistribution(distId),
    [qk.documentDistributions(docId)],
  );
}

// ---------- categories CRUD (admin-ui slice 04,先落 hook) ----------
// knowledge-tiered admin-ui:Category write hooks。成功后失效 categories 列表。

/** 新建本级 Category(scope 按 role 校验在后端 service 层)。 */
export function useCreateCategory() {
  return useApiMutation(
    (payload: KnowledgeCategoryCreate) => createCategory(payload),
    [qk.knowledgeCategories],
  );
}

/** 改 Category 的 name/sort_order(scope 不可改,对齐后端 KnowledgeCategoryUpdate)。 */
export function useUpdateCategory() {
  return useApiMutation(
    ({ id, payload }: { id: string; payload: KnowledgeCategoryUpdate }) =>
      updateCategory(id, payload),
    [qk.knowledgeCategories],
  );
}

/** 软删 Category(name 释放可复用)。 */
export function useDeleteCategory() {
  return useApiMutation(
    (id: string) => deleteCategory(id),
    [qk.knowledgeCategories],
  );
}

